test/spl/degree_reduction_test.cc

```diff
@@ -0,0 +1,384 @@
+/* Copyright 2018 Chair for Computational Analysis of Technical Systems, RWTH Aachen University
+
+This file is part of SplineLib.
+
+SplineLib is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public
+License as published by the Free Software Foundation version 3 of the License.
+
+SplineLib is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along with SplineLib.  If not, see
+<http://www.gnu.org/licenses/>.
+*/
+
+#include <memory>
+
+#include "gmock/gmock.h"
+
+#include "b_spline.h"
+#include "nurbs.h"
+#include "random_b_spline_generator.h"
+#include "random_nurbs_generator.h"
+
+using testing::Test;
+
+class BSplineFig5_35ForDegreeElevationAndReductionForDimension0 : public Test {  // NOLINT
+ public:
+  BSplineFig5_35ForDegreeElevationAndReductionForDimension0() {
+    std::array<Degree, 1> degree = {Degree{3}};
+    KnotVectors<1> knot_vector_before = {std::make_shared<baf::KnotVector>(
+        baf::KnotVector({ParamCoord{0}, ParamCoord{0}, ParamCoord{0}, ParamCoord{0}, ParamCoord{0.3}, ParamCoord{0.6},
+                         ParamCoord{1}, ParamCoord{1}, ParamCoord{1}, ParamCoord{1}}))};
+    std::vector<baf::ControlPoint> control_points = {
+        baf::ControlPoint(std::vector<double>({1.0, 0.0})),
+        baf::ControlPoint(std::vector<double>({0.0, 2.0})),
+        baf::ControlPoint(std::vector<double>({1.0, 2.0})),
+        baf::ControlPoint(std::vector<double>({3.0, 2.0})),
+        baf::ControlPoint(std::vector<double>({4.0, 1.0})),
+        baf::ControlPoint(std::vector<double>({3.0, 0.0}))
+    };
+    original_ = std::make_shared<spl::BSpline<1>>(knot_vector_before, degree, control_points);
+    elevated_and_reduced_ = std::make_shared<spl::BSpline<1>>(*original_);
+    elevated_and_reduced_->ElevateDegreeForDimension(0);
+    successful_ = elevated_and_reduced_->ReduceDegreeForDimension(0);
+  }
+
+ protected:
+  std::shared_ptr<spl::BSpline<1>> original_;
+  std::shared_ptr<spl::BSpline<1>> elevated_and_reduced_;
+  bool successful_;
+};
+
+TEST_F(BSplineFig5_35ForDegreeElevationAndReductionForDimension0, // NOLINT
+    DegreeReductionWithoutPreviousDegreeElevationFails) { // NOLINT
+  std::shared_ptr<spl::BSpline<1>> test_spline = std::make_shared<spl::BSpline<1>>(*original_);
+  bool successful = test_spline->ReduceDegreeForDimension(0);
+  ASSERT_THAT(successful, false);
+  ASSERT_THAT(test_spline->GetDegree(0), original_->GetDegree(0));
+  ASSERT_THAT(test_spline->GetKnotVector(0)->GetNumberOfKnots(), original_->GetKnotVector(0)->GetNumberOfKnots());
+  ASSERT_THAT(test_spline->GetNumberOfControlPoints(), original_->GetNumberOfControlPoints());
```

Comment:
Use AreEqual instead of these 3 tests as AreEqual tests exactly the same and even more.